crates/rpc-types-beacon/src/payload.rs

```diff
@@ -435,9 +435,71 @@ pub mod beacon_payload_v3 {
     }
 }
 
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize)]
+#[cfg(feature = "amsterdam")]
+struct BeaconExecutionPayloadV4<'a> {
+    /// Inner V3 payload
+    #[serde(flatten)]
+    payload_inner: BeaconExecutionPayloadV3<'a>,
+    /// RLP encoded `block_access_list`
+    block_access_list: Cow<'a, Bytes>,
+}
+
+#[cfg(feature = "amsterdam")]
+impl<'a> From<BeaconExecutionPayloadV4<'a>> for alloy_rpc_types_engine::ExecutionPayloadV4 {
+    fn from(payload: BeaconExecutionPayloadV4<'a>) -> Self {
+        let BeaconExecutionPayloadV4 { payload_inner, block_access_list } = payload;
+        Self {
+            payload_inner: payload_inner.into(),
+            block_access_list: block_access_list.into_owned(),
+        }
+    }
+}
+
+#[cfg(feature = "amsterdam")]
+impl<'a> From<&'a alloy_rpc_types_engine::ExecutionPayloadV4> for BeaconExecutionPayloadV4<'a> {
+    fn from(value: &'a alloy_rpc_types_engine::ExecutionPayloadV4) -> Self {
+        let alloy_rpc_types_engine::ExecutionPayloadV4 { payload_inner, block_access_list } = value;
+        BeaconExecutionPayloadV4 {
+            payload_inner: payload_inner.into(),
+            block_access_list: Cow::Borrowed(block_access_list),
+        }
+    }
+}
+
+/// A helper serde module to convert from/to the Beacon API which uses quoted decimals rather than
+/// big-endian hex.
+#[cfg(feature = "amsterdam")]
+pub mod beacon_payload_v4 {
+    use super::*;
+
+    /// Serialize the payload attributes for the beacon API.
+    pub fn serialize<S>(
+        payload_attributes: &alloy_rpc_types_engine::ExecutionPayloadV4,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        BeaconExecutionPayloadV4::from(payload_attributes).serialize(serializer)
+    }
+
+    /// Deserialize the payload attributes for the beacon API.
+    pub fn deserialize<'de, D>(
+        deserializer: D,
+    ) -> Result<alloy_rpc_types_engine::ExecutionPayloadV4, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        BeaconExecutionPayloadV4::deserialize(deserializer).map(Into::into)
+    }
+}
+
 /// Represents all possible payload versions.
 #[derive(Debug, Serialize)]
 #[serde(untagged)]
+#[cfg(not(feature = "amsterdam"))]
```

Comment:
instead of feature gating the type, can we instead feature gate the variant instead?